### PR TITLE
AB#473 Feature/us279: demo fw for api testing

### PR DESCRIPTION
**Pull request by svchig**
add  API framework

Created a simple test automation framework for API testing with BaseScenario:
1. find any product (POST products)
2. add found product to cart (POST cartLines/)
3. check that product exists in the cart (GET cat)
4. check deleveryInfo, shippingInfo, billibgInfo (GET)
5. add shipping info (POST shippingInfo)
6. add graphql ClientConfiguration (POST query), add graphql TokenizeCreditCard (POST mutation)
7. add paymentInfo (POST)
8. order to pay (POST)
9. check order request by confirmationID (GET)
10. Check that cart is empty (GET cart)

Note: to run API BaseScenario test, need to add BraintreeMerchantId, BraintreePublicKey, BraintreePrivateKey values into App.config

### CODE REVIEW

**Comment by svchig** [comment]
need to replace $(hca_hostname), $(sitecore_analytics_clobal_cookie), $(braintree_sendbox_url), $(braintre_merchant_id), $(braintre_public_key), $(braintre_private_key), $(braintree_version) values in App.config of 'HcaApiTests' project before build it

File: tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config [hunk 1]
```diff
@@ -0,0 +1,26 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <appSettings>
+    <add key="Host" value="$(hca_hostname)" />
+    <add key="BaseUrl" value="https://$(Host)/apix/client/commerce" />
+    <add key="BaseSCUrl" value="https://$(Host)/sitecore/api/layout/render" />
+    <!-- use '$(sitecore_analytics_clobal_cookie)' e.g. "b916df89e9444a968dc1b74f53813e1e|True"-->
```
Comment (larionovadaria) [inline tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config:7]: clobal => global

Please choose one type of naming: 
`Braintree_Merchant_Id`
`BraintreePublicKey`
`SC_ANALYTICS_GLOBAL_COOKIE`

File: tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    <add key="SC_ANALYTICS_GLOBAL_COOKIE" value="$(sitecore_analytics_clobal_cookie)" />
+    <!-- use '$(braintree_sendbox_url)' e.g. "https://payments.sandbox.braintree-api.com/graphql"-->
+    <add key="GraphQLSendboxUrl" value="$(braintree_sendbox_url)" />
+    <add key="Braintree_Merchant_Id" value="$(braintre_merchant_id)" />
```
Comment (larionovadaria) [inline tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config:11]: here and below: braintre_ => braintree_

File: tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <appSettings>
+    <add key="Host" value="$(hca_hostname)" />
+    <add key="BaseUrl" value="https://$(Host)/apix/client/commerce" />
+    <add key="BaseSCUrl" value="https://$(Host)/sitecore/api/layout/render" />
+    <!-- use '$(sitecore_analytics_clobal_cookie)' e.g. "b916df89e9444a968dc1b74f53813e1e|True"-->
+    <add key="SC_ANALYTICS_GLOBAL_COOKIE" value="$(sitecore_analytics_clobal_cookie)" />
+    <!-- use '$(braintree_sendbox_url)' e.g. "https://payments.sandbox.braintree-api.com/graphql"-->
+    <add key="GraphQLSendboxUrl" value="$(braintree_sendbox_url)" />
```
Comment (larionovadaria) [inline tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config:10]: sendbox => sandbox

File: tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <appSettings>
+    <add key="Host" value="$(hca_hostname)" />
+    <add key="BaseUrl" value="https://$(Host)/apix/client/commerce" />
+    <add key="BaseSCUrl" value="https://$(Host)/sitecore/api/layout/render" />
+    <!-- use '$(sitecore_analytics_global_cookie)' e.g. "b916df89e9444a968dc1b74f53813e1e|True"-->
```
Comment (larionovadaria) [inline tests/API/HcaApiTestAutomationFramework/HcaApiTests/App.config:7]: Remove comments

File: tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/GraphQlClient.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+			gateway.Configuration.Timeout = 10000;
+			var clientToken = gateway.ClientToken.Generate();
+			return clientToken;
+		}
+
+		public string Base64Decoder(string stringText)
+		{
+			string base64Encoded = stringText;
+			string base64Decoded;
+			byte[] data = System.Convert.FromBase64String(base64Encoded);
```
Comment (VitaminkaA) [inline tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/GraphQlClient.cs:33]:
```suggestion
			var base64Encoded = stringText;
			var base64Decoded;
			var data = System.Convert.FromBase64String(base64Encoded);
```
why do you use explicit type declaration in variables?
- var requires less typing. var is shorter and easier to read, for instance, than Dictionary <int, IList <string>>.
- var requires less code changes if the return type of a method call changes. You only have to change the method declaration, not every place it’s used.
- var encourages you to use a descriptive name for the variable. This means the instance, not the type name. For instance:
- var customer = new Customer () rather than var c = new Customer ().

>

File: tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/GraphQlDTO/ClientConfigurationRequestDTO.cs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+﻿namespace HcaApiTestAutomationFramework.GraphQlDTO
+{
+	public class ClientConfigurationRequestDTO
+	{
+		public Clientsdkmetadata clientSdkMetadata { get; set; }
+		public string query { get; set; }
+		public string operationName { get; set; }
```
Comment (VitaminkaA) [inline tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/GraphQlDTO/ClientConfigurationRequestDTO.cs:7]: The names of all public fields, properties must begin with a capital letter
```suggestion
		public ClientSdkMetadata СlientSdkMetadata { get; set; }
		public string Query { get; set; }
		public string OperationName { get; set; }
```

File: tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/HcaApiMethods.cs [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+			var request = cart.CreateGetRequest();
+			var response = cart.GetResponse(url, request);
+			CartResponseDTO content = cart.GetContent<CartResponseDTO>(response);
+			return content;
+		}
+
+		public OrdersOrderResponseDTO GetOrder(string confirmationId)
+		{
+			var cart = new ApiHelper<OrdersOrderResponseDTO>();
+			var endpoint = "orders/" + confirmationId;
```
Comment (VitaminkaA) [inline tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/HcaApiMethods.cs:32]:
```suggestion
			var endpoint = $"orders/{ confirmationId}";
```

File: tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/GraphQlClient.cs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+			AddRequestHeaders(request);
+			request.Timeout = timeout;
+
+			if (additionalHeaders != null && additionalHeaders.Count > 0)
+			{
+				foreach (var additionalHeader in additionalHeaders)
+				{
+					request.AddHeader(additionalHeader.Key, additionalHeader.Value);
+				}
+			}
```
[the remark this answers is not in the document]
Comment (evgenii-nikulin) [reply tests/API/HcaApiTestAutomationFramework/HcaApiTestAutomationFramework/GraphQlClient.cs:71]: You can use `Resharper` for cleanup. In development part we use brackets everywhere for more readability.

File: src/Project/HCA/code/App_Config/Include/Project/HCA/Project.HCA.Variables.config [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿<configuration>
+  <sitecore>
+    <sc.variable name="HCA.HostName" value="" />
```
Comment (larionovadaria) [inline src/Project/HCA/code/App_Config/Include/Project/HCA/Project.HCA.Variables.config:3]: Why did you add this one?

Comment (larionovadaria) [reply src/Project/HCA/code/App_Config/Include/Project/HCA/Project.HCA.Variables.config:3]: And `BIN +813 KB Homepage.png ` above

**Comment by svchig** [comment]
too many conflicts
